---
title: "Class 15: Pathway Analysis from RNA-Seq Results"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only.


Read in the data
```{r}
countData <- read.csv("GSE37704_featurecounts.csv", row.names = 1)
colData <- read.csv("GSE37704_metadata.csv", row.names = 1)

head(countData)
head(colData)
```

Remove the length columun from countData because it is not needed
```{r}
#Don't change the data and then put it in the same variable because it overwrites the data
countData <- countData[,-1]
head(countData)
```

Need to remove the genes that have 0s for every experiment 
```{r}
countData <- countData[rowSums(countData) > 0,]
nrow(countData)
```

15975 genes is alot to plot so let's use principal componet analysis on the data to see if the controls and experiment are different
```{r}
#t = transpose, prcomp needs the data to be in columns
pc <- prcomp(t(countData))
plot(pc)
```

```{r}
summary(pc)
```

```{r}
mycols = c(rep("blue", 3), rep("red", 3))
plot(pc$x[,1:2], col = mycols)
```

##Differential expression analysis
```{r}
library(DESeq2)
```

```{r}
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData,
                              design = ~condition)
dds <- DESeq(dds)
```

```{r}
dds
```
```{r}
res <- results(dds)
summary(res)
```

##Volcano plot summary
```{r}
plot( res$log2FoldChange, -log(res$padj) )
```

#Add color
```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```


#Add gene annotations
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)
```

```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys= row.names(res),  #where are your gene IDs?
                    keytype="ENSEMBL", #what format are your gene IDs?
                    column= "SYMBOL",   #what format do you want the names changed to?   
                    multiVals="first")  #If there are more than one name, which one do you pick?

res$entrez = mapIds(org.Hs.eg.db,
                    keys= row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

```

```{r}
head(res, 10)
```

##Pathway analysis
```{r}
# Run in your R console (i.e. not your Rmarkdown doc!)
#BiocManager::install( c("pathview", "gage", "gageData") )
library(pathview)
library(gage)
library(gageData)
```


```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

The main gage() function requires a named vector of fold changes, where the names of the values are the Entrez gene IDs.

Note that we used the mapIDs() function above to obtain Entrez gene IDs (stored in res$entrez) and we have the fold change results from DESeq2 analysis (stored in res$log2FoldChange).
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
attributes(keggres)
```

Show all the pathways that are down regulated
```{r}
head(keggres$less)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![Pathway graphic](./hsa04110.pathview.png)








